hubblestack/files/hubblestack_nova/oval_scanner.py

```diff
@@ -166,23 +174,128 @@ def build_impact_report(impact, report={}):
+    """Compare local package version to vulnerability version"""
+    logging.debug('impact_worker')
     impact = {}
-    if __salt__['pkg.version_cmp'](ver, local_ver) > 0:
-        impact[title] = {
-            'updated_pkg': {'name': name, 'version': ver},
-            'installed': {'name': name, 'version': local_ver},
-            'severity': severity,
-            'advisory': advisory,
-            'cve': cve
-        }
+    if distro_name in ('centos', 'redhat'):
+        impact = parse_rpm_version(impact, local_ver, **kargs)
+    elif __salt__['pkg.version_cmp'](kargs['ver'], local_ver) > 0:
+        impact = create_impact(impact, local_ver, **kargs)
     return impact
 
 
+def parse_rpm_version(impact, local_ver, **kargs):
+    """rpmdevtools ver comparison is broken, so manually compare versions"""
+    logging.debug('parse_rpm_version')
+    local_version = local_ver.split('.centos')[0]
+    if not re.search(":", local_version):
+        local_version = '0:' + local_version
+    split_local_ver = re.split(':+|.el', local_version)
+    split_impact_ver = re.split(':+|.el', kargs['ver'])
+    if (
+        not re.search('.el', local_ver, re.IGNORECASE)
+        and parse_version(split_local_ver[0]) <= parse_version(split_impact_ver[0])
+        and parse_version(split_local_ver[1]) < parse_version(split_impact_ver[1])
+    ) or (
+        parse_version(split_local_ver[0]) <= parse_version(split_impact_ver[0])
+        and parse_version(split_local_ver[1]) < parse_version(split_impact_ver[1])
+        and split_local_ver[2] <= split_impact_ver[2]
+    ):
+        impact = create_impact(impact, local_ver, **kargs)
+    return impact
+
+
+def create_impact(impact, local_ver, **kargs):
+    impact[kargs['title']] = {
+        'updated_pkg': {'name': kargs['name'], 'version': kargs['ver']},
+        'installed': {'name': kargs['name'], 'version': local_ver},
+        'severity': kargs['severity'],
+        'advisory': kargs['advisory'],
+        'cve': kargs['cve']
+    }
+    return impact
+
+
+# Build source package reference to binary packages
+def build_pkg_src_ref(local_pkgs, queue, pkg_src_ref={}):
+    """Build source package reference for binary packages"""
+    logging.debug('build_pkg_src_ref')
+    get_package_detail(local_pkgs, queue)
+    while not queue.empty():
+        pkg_detail = queue.get()
+        for pkg, detail in pkg_detail.items():
+            if 'Source' in next(iter(detail.values())):
+                source = next(iter(detail.values()))['Source'].split()[0]
+                if source not in pkg_src_ref and not source == pkg:
+                    pkg_src_ref[source] = []
+                if not source == pkg:
+                    pkg_src_ref[source].append(pkg)
+        queue.task_done()
+    return pkg_src_ref
+
+
+def get_package_detail(local_pkgs, queue):
+    """Get package details"""
+    threads = []
+    for pkg in local_pkgs:
+        worker = threading.Thread(target=pkg_show, args=(pkg, queue,))
```

Comment:
A big NO. This is not recommended.

On a fresh VM, I got around 1450 packages. Which means, you are launching 1450 threads (will vary for each machine).
This will kill performance and memory.

Better to use some thread-pool with much lesser number of threads.

Reply:
Also, should give some timeout value as well. You never know when a thread becomes unresponsive or reached in deadlock state. 
And since there is almost no logging, we will have a hard time in debugging such issues.

Reply:
I'd really rather not use any threading at all. Hubble is single threaded on purpose. We could establish a worker thread pool or something, but we don't because we don't ever want Hubble to be something that consumes enough resources for admins to notice.

I would say loading threading at all is a good reason to not merge this PR.